### DISCUSSION TITLE
Custom Identity-Provider properties are not shown

### DISCUSSION THREAD

**Discussion by klausbetz** [keycloak/keycloak#11849] [Archive]
Having a sneak peak on the fancy new `admin2 UI` I found that the custom theme (which contains form fields for custom properties) from this identity provider is ignored.
[https://github.com/klausbetz/apple-identity-provider-keycloak/blob/main/src/main/resources/theme-resources/resources/partials/realm-identity-provider-apple.htm](https://github.com/klausbetz/apple-identity-provider-keycloak/blob/main/src/main/resources/theme-resources/resources/partials/realm-identity-provider-apple.html)

Is there a way to tell the new UI to display the custom attributes of the IDP, besides the standard ones like Client-ID?
`Service ID`, `Team ID`, `Key ID` and so on are missing.

Working example in old UI:
<img width="1792" alt="Bildschirmfoto 2022-05-05 um 15 36 18" src="https://user-images.githubusercontent.com/78362353/166935408-d2bf6c98-8077-4fca-8d2c-4ee1d206b9da.png">

New UI:
<img width="1792" alt="Bildschirmfoto 2022-05-05 um 15 34 39" src="https://user-images.githubusercontent.com/78362353/166935393-512eb21e-da50-44eb-9d31-668538bac99a.png">

**Comment by klausbetz** [comment]
@stianst, @edewit: Do you have any idea how this can be achieved? I'd be more than happy to offer a new version of the Apple IDP that is compatible to the new Admin UI 😄

**Reply by edewit** [reply]
@klausbetz right the new ui is no longer based on angular so this will not work it. ATM we don't support any plugins for the ui other then the SPIs that where there before as well. The way to do this now is by forking the ui project and then building it in.

**Reply by lscorcia** [reply]
See also https://github.com/keycloak/keycloak-ui/issues/3696 . This is quite important and I am not entirely sure it is meant to be this way since a similar issue has been fixed for custom user storage (https://github.com/keycloak/keycloak-ui/issues/3059).

**Reply by edewit** [reply]
@lscorcia this issue is not the same as the issue you mentioned, because in this issue it's about extending the admin ui with your own page

**Reply by klausbetz** [reply]
I think the initial description of _custom theme is ignored_ is misleading here.  
Basically, the issue is almost identical to the one @lscorcia mentioned. Let me explain this one further.

The old admin UI required Identity-Provider extension developers to add a custom page with all their custom properties in there, in order to have them shown on the respective IDP configuration page.  
In the end it's about custom properties to be shown (not a custom theme or page) and this was only possible using a custom page.  

I'd be more than happy to skip the custom page/theme part and just focus on the custom properties that really matter (`Service ID`, `Team ID`, `Key ID` and `p8 Key`).

It's great to have the possibility of forking and extending the new ui project.  
But how are extension developers supposed to maintain and ship their own admin theme?

**Reply by stianst** [reply]
It must be possible to configure custom identity providers in the new admin console without the need to fork the project.

I thought the identity providers use the component model stuff so has the ability to define dynamic configuration options that the admin console should support similarly to protocol mappers for example.

**Reply by edewit** [reply]
yes it does, but this is different apparently we used to support just adding a html file that is then rendered as part of the ui. The new ui isn't based on html so we can't really support this anymore.

**Comment by stianst** [comment]
Looking at this quickly IdentityProviders are not configurable it seems:

https://github.com/keycloak/keycloak/blob/f616495b051ae9555fe97df0e269c8332401b8d5/server-spi-private/src/main/java/org/keycloak/broker/provider/IdentityProviderFactory.java#L29

Take a look at protocol mappers for example they are:

https://github.com/keycloak/keycloak/blob/f616495b051ae9555fe97df0e269c8332401b8d5/server-spi-private/src/main/java/org/keycloak/protocol/ProtocolMapper.java#L32

So, this is something that was completely missed for the new admin console, and it seems it really isn't possible to make identity providers configurable properly in the new admin console, and we'll have to refactor the identity providers to ConfiguredProvides: 

https://github.com/keycloak/keycloak/blob/f616495b051ae9555fe97df0e269c8332401b8d5/server-spi/src/main/java/org/keycloak/provider/ConfiguredProvider.java#L29

**Reply by lscorcia** [reply]
Just wanted to add a detail - while having the ability to edit the configuration values would be an acceptable minimum feature, the previous customization model allowed us to express dependencies between configuration values. For example in our plugin a whole configuration section was enabled or disabled conditionally according to some other parameter value. Also, we added a link to an additional SAML metadata document. AFAICS this doesn't seem to be possible with the "configurable" pattern.

**Reply by edewit** [reply]
yep that is true it does allow for more flexibility

**Reply by stianst** [reply]
It could be possible perhaps to extend configured providers to allow stuff like that (some sort of conditions on a property), but let's start somewhere by leveraging what we already have around this.

**Comment by stianst** [comment]
Created a follow-up issue here:

https://github.com/keycloak/keycloak/issues/15344

Ability to create custom identity providers, and make them configurable through the admin console is something we have supported for a long time. Not having a solution here basically blocks us from being able to remove the old admin console, as the new admin console doesn't have all the features required.
